### ISSUE TITLE
Performance question re: large directories

### ISSUE THREAD

**Issue by toptensoftware** [fabiospampinato/watcher#16]
I'm wondering how well this library handles large directories.  

Use case:

- 100,000+ files
- Not deep, maybe 5-7
- Linux 
- Infrequent updates (days to weeks apart)
- Changes are usually new files (maybe 100)

Is this feasible? In particular I want to avoid excessive memory usage and periodic CPU load spikes. Less cpu/memory would be preferable to timely updates.

**Comment by fabiospampinato** [comment]
I can't say that I've tested that use case myself, but if I had to guess:

- There's no native filesystem watcher for watching folders deeply for Linux provided by Node, so the library will watch every single folder and file by itself. I've found this to be better on average actually because it allows you to ignore some specific paths, and watching huge unintended folders (like node_modules or something) is where I've seen most of the performance issues happening.
- The number of files shouldn't be an issue. Assuming the process has access to enough file descriptors I guess.
- The number of folders maaaybe might be, somebody reported an issue on another library I internally use (https://github.com/fabiospampinato/promise-concurrency-limiter/pull/1) that I guess could potentially cause Watcher to not work, but I think it's like either it doesn't work at all or if you see it working it's not a problem for your use case. I wanted to try to address that today actually. I think I just need to read folders depth-first rather than breath-first or something in this other library (https://github.com/fabiospampinato/tiny-readdir).
- CPU-wise as long as you are watching paths that exist there shouldn't be any polling happening, so I guess if no events happen for days basically no CPU should be used for days. 
- I was thinking the longer the process is kept running the longer there's a chance that any eventual internal memory leak will blow out of proportions, but if there's nothing happening there are also no allocations happening, and the number of events seems small so even if there were memory leaks they probably wouldn't matter for a long time.
- Memory-wise it will consume less memory than other watchers I believe, including watching files normally with Node itself. [Internally](https://github.com/fabiospampinato/watcher/blob/master/src/watcher_stats.ts) I'm using a more efficient data structure for keeping track of stats objects, that helps with keeping memory usage low. Node by default gives you a million useless Date objects that you can construct for yourself whenever you want it, they both waste time for being created and waste memory for being kept in memory. I kinda overdid this part of the project, reading and storing in memory stats objects is like 2x faster than what you can do with Node itself (https://github.com/fabiospampinato/ripstat/blob/db5fe5b4870c093889b29cd44e1dcdef722830b6/src/index.ts#L11), and I'm doing this within Node 🤣

Basically I think it should work just fine and memory usage should be lower compared to other JS-based filesystem watchers. 

If you end up using this it'd be interesting if you could share some findings.

**Comment by toptensoftware** [comment]
Thanks for the fast, detailed reply. I'll let you know what I find if I go ahead with it.
